rfd/0073-idp-initiated-login.md

```diff
@@ -0,0 +1,67 @@
+
+OIDC does not have a commonly used mechanism for IdP-initiated login. This has resulted in IdP support for this feature being scarce and limited. Furthermore, the Third Party Initiated Login feature of OIDC is not well specified in the standard as it is generally used to create a middleman IdP that sits between between a true SP and IdP using a completely custom flow between an SP and an IdP which you both control. Therefore OIDC lacks the tooling and product support from IdP's to make this feature useful.
+
+For this reason, the rest of the RFD focuses purely on SAML and assumes that is the only connector we will support IdP-initiated login for.
+
+### Configuration
+
+Add a new configurable field to SAML connectors called `allow_idp_initiated` which defaults to false. If set to true, Teleport will accept the SAML assertion sent by the IdP even if it is not requested by Teleport. This allows you to configure the SAML callback in an IdP provider as usual and log in to Teleport directly from the IdP dashboard.
+
+```
+kind: saml
+version: v2
+metadata:
+  name: new_saml_connector
+spec:
+  display: MyIdP
++ allow_idp_initiated: true
+  # other fields
+```
+
+### Security Considerations
+
+Allowing IdP-initiated login flows comes with a set of security tradeoffs inherent to the use of SAML and the flow itself. In this configuration, Teleport becomes more prone to replay and CSRF attacks since Teleport cannot verify that the user started the login flow intentionally. This opens up attack vectors where an attacker can trick a legitimate user into unknowingly logging into the application with the identity of the attacker.
+
+Due to the security risks below, this feature should be opt-in using the configuration option above and not enabled by default. Since this can reduce usage friction, we should still offer the feature to those that want it.
+
+#### Specific risks
+
+- Replay Attacks: Replay attacks are possible by capturing SAML assertions and possibly reusing them. We can attempt to mitigate this in Teleport by storing a cache of SAML assertion IDs that are valid for as long the SAML assertion is itself valid. This may not always be feasible depending on the data we receive from the IdP, but it is an useful mitigation we should employ. 
+
+- SAML interception: Since no sort of nonce or request/response system can be employed with randomly generated CSRF relay state tokens or other mechanism, IdP-initated is prone to being chained with other attacks using assertion injection, interception and stealing and reuse by an attacker. We can attempt to mitigate this by reducing the trust length (the time we consider a response to be valid for login from it's issuance) but this is ultimately just a mitigation and not a perfect solution. We should also check the `InResponseTo` field to ensure any stolen responses aren't reused with the IdP-initated flow.
```

Comment:
```suggestion
- SAML interception: Since no sort of nonce or request/response system can be employed with randomly generated CSRF relay state tokens or other mechanism, IdP-initiated is prone to being chained with other attacks using assertion injection, interception and stealing and reuse by an attacker. We can attempt to mitigate this by reducing the trust length (the time we consider a response to be valid for login from it's issuance) but this is ultimately just a mitigation and not a perfect solution. We should also check the `InResponseTo` field to ensure any stolen responses aren't reused with the IdP-initated flow.
```